src/utils/EIP6963Provider.class.ts

```diff
@@ -0,0 +1,64 @@
+import { EIP1193Provider, EIP6963ProviderDetail } from './types.ts'
+import { EIP6963_MANAGER_INJECTED_PROVIDERS_MAP } from '../providers/EIP6963ManagerContext.ts'
+
+interface Listener<T = unknown> {
+  (...args: T[]): void
+}
+
+export class EIP6963ClassProvider implements EIP1193Provider {
+  currentProviderDetail?: EIP6963ProviderDetail
+
+  // Stores stable references to prevent memory leaks
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private readonly proxyListeners: { [eventName: string | symbol]: Listener<any>[] } = {}
+
+  request = async <T = object | unknown[], R = unknown>(args: { method: string; params?: T }): Promise<R> =>
+    this.currentProviderDetail && this.currentProviderDetail.provider
+      ? this.currentProviderDetail.provider.request<T, R>(args)
+      : Promise.reject('Provider is undefined')
+
+  on = <T = unknown>(eventName: string, listener: Listener<T>): this => {
+    if (!this.proxyListeners[eventName]) {
+      this.proxyListeners[eventName] = []
+    }
+    this.proxyListeners[eventName].push(listener)
+    this.currentProviderDetail?.provider.on(eventName, listener)
+
+    return this
+  }
+
+  removeListener = <T = unknown>(eventName: string | symbol, listener: Listener<T>): this => {
+    this.currentProviderDetail?.provider.removeListener(eventName, listener)
+
+    if (this.proxyListeners[eventName]) {
+      const index = this.proxyListeners[eventName]?.indexOf(listener)
```

Comment:
```suggestion
      const index = this.proxyListeners[eventName]?.lastIndexOf(listener)
```

---

I tried adding duplicate listeners on MetaMask and Brave and they remove the last instance:
```js
const providers = []
window.addEventListener("eip6963:announceProvider", (event) => {
  providers.push(event.detail);
});
window.dispatchEvent(new Event("eip6963:requestProvider"));

const listener = (e) => console.log('listener')
providers[0].provider.on('chainChanged', () => console.log(1))
providers[0].provider.on('chainChanged', listener)
providers[0].provider.on('chainChanged', () => console.log(2))
providers[0].provider.on('chainChanged', listener)
providers[0].provider.on('chainChanged', () => console.log(3))
providers[0].provider.on('chainChanged', listener)
providers[0].provider.on('chainChanged', () => console.log(4))

await providers[0].provider.request({ method: 'eth_chainId' })
// switch chain
// > 1, listener, 2, listener, 3, listener, 4

providers[0].provider.removeListener('chainChanged', listener)
// switch chain
// > 1, listener, 2, listener, 3, 4
```

Reply:
So you added 7 `chainChanged` listeners, which is highly unlikely scenario. The current implementation will deny any additional listeners - so that it ensures that each listener is registered only once:
https://github.com/oasisprotocol/dapp-wrose/blob/70912a8b5eda8cc27cb4ddd0c802c8786be617a0/src/providers/Web3Provider.tsx#L101-L115

Reply:
Oh, more bugs: _addEventListenersOnce is created many times. If you call `connectWallet(ProviderType.EIP6963)` again later it will attach more listeners.

Reply:
Sure, but this happens only once in the current app, but true, `_addEventListenersOnce` needs to be wrapped inside useCallback, as it will return new instance every time. Or move `eventListenersInitialized` outside the ContextProvider, which makes more sense I guess.